packages/core/src/api.ts

```diff
@@ -175,15 +194,17 @@ export const deploy = async (
   }
 
   const postDeployAction = async (appliedChanges: ReadonlyArray<Change>): Promise<void> => {
-    await promises.array.series(appliedChanges.map(change => async () => {
-      const updatedElement = await getUpdatedElement(change)
-      if (change.action === 'remove' && !isFieldChange(change)) {
-        await workspace.state().remove(updatedElement.elemID)
-      } else {
-        await workspace.state().set(updatedElement)
-        await changedElements.set(updatedElement)
-      }
-    }))
+    // create updated elements before updating the state, to prevent any unwanted behavior
+    const updatedElements = awu(appliedChanges)
+      .filter(change => (isAdditionOrModificationChange(change) || isFieldChange(change)))
+      .map(getUpdatedElement)
+
+    const detailedChanges = appliedChanges.map(change => getDetailedChangesFromChanges(change)).flat()
+    await workspace.state().updateStateFromChanges({ changes: detailedChanges })
+
+    await updatedElements.forEach(async updatedElement => {
```

Comment:
```suggestion
    // update changedElements before updating the state because we rely on the state data for the calculation
    await awu(appliedChanges)
      .filter(change => (isAdditionOrModificationChange(change) || isFieldChange(change)))
      .map(getUpdatedElement)
      .forEach(async updatedElement => {
        await changedElements.set(updatedElement)
      })

    const detailedChanges = appliedChanges.flatMap(getDetailedChangesFromChanges)
    await workspace.state().updateStateFromChanges({ changes: detailedChanges })
```

as an alternative, we could keep just the IDs of the changed elements and use the state as the element source for comparison in [line 218](https://github.com/salto-io/salto/pull/4227/files#diff-7a68aa0c5146864913c9bb8b990a25deb8715ead9aea7f9aa9e0f44d23d57448R218)

Reply:
That means that 'adaptersElementSource' won't have changedElements as a a fallback source
Im not sure in what case changedElement will include elements that that aren't in workspace.elements(), but I don't have enough knowledge to be sure

Reply:
ok, so not the alternative then